[nativert] Move Graph to pytorch core

yiming0416 opened the pull request:
Summary: 
Torch Native Runtime RFC: https://github.com/pytorch/rfcs/pull/72

4 classes have been introduced: `Graph`, `Node`, `Value`, `Type`
- `Type` represents the kind of a `Value`
- `Value` represents a single symbolic value, it could be any kind that exists in `Type`. Values are inputs and outputs of a `Node`. 
- `Node` represents a single unit of execution, typically a PyTorch op. 
- `Graph` represents a model's computation graph, which is designed to facilitate transformation/analysis. 

Test Plan: Added test under `test/cpp/nativert/test_graph.cpp`

Differential Revision: D74749418

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+  return v;
+}
+
+void Graph::addConstantOutput(Constant& c) {
```
```suggestion
void Graph::addConstantOutput(Constant&& c) {
```
or even
```suggestion
void Graph::addConstantOutput(Constant c) {
```
curious you are moving an lvalue here?

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+    std::string target,
+    std::vector<NamedArgument> inputs,
+    std::unordered_map<std::string, std::string> metadata) {
+  auto node = std::make_unique<Node>(
+      this, std::move(target), std::move(inputs), std::move(metadata));
+  auto ret = node.get();
+  nodesOwner_.push_back(std::move(node));
+  return ret;
```
```suggestion
  auto & node = nodesOwner_.emplace_back(std::make_unique<Node>(
      this, std::move(target), std::move(inputs), std::move(metadata)));
  return node.get();
```

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+}
+
+Node* Graph::createListPack(std::vector<Value*> inputs, Type inputType) {
+  std::vector<NamedArgument> nodeInputs;
```
```suggestion
  std::vector<NamedArgument> nodeInputs;
  nodeInputs.reserve(inputs.size());
```

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+}
+
+Node* Graph::createOptionalListPack(std::vector<Value*> inputs) {
+  std::vector<NamedArgument> nodeInputs;
```
```suggestion
  std::vector<NamedArgument> nodeInputs;
  nodeInputs.reserve(inputs.size());

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+    }
+  }
+
+  std::vector<Value*> outputValues;
+  for (auto* outputValue : subgraph.outputs()) {
```
```suggestion
  auto outputs = subgraph.outputs();
  std::vector<Value*> outputValues;
  outputValues.reserve(outputs.size());
  for (auto* outputValue : outputs()) {
```

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+    std::string target,
+    std::vector<NamedArgument> inputs,
+    std::unordered_map<std::string, std::string> metadata)
+    : owningGraph_(owningGraph),
```
why a ptr instead of a reference, can owningGraph be null?

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+
+std::ostream& operator<<(std::ostream& out, const Graph& graph) {
+  for (const auto& node : graph.nodes_) {
+    out << node << "\n";
```
You want tthis to end with an extra newline?

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+}
+
+bool Parser::validIdent(char n) {
+  return isalpha(n) || n == '_' || isdigit(n);
```
```suggestion
  return std::isalpha(n) || n == '_' || std::isdigit(n);

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+  return ret;
+}
+
+bool Parser::validIdent(char n) {
```
Beware of signed char, most of these helpers only take unsigned char

Skylion007 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+}
+
+std::string graphToString(const Graph& g, bool include_signature) {
+  std::stringstream ss;
```
fmtlib might be a bit faster, but I can see if you don't want to setup a custom formatter for libfmt

yiming0416 replied:
This is function is only used in tests, so I guess it's fine.

Skylion007 reviewed (commented):
lots of performance nits here

yiming0416 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+  }
+
+  // Sort values in creation order (by value ids)
+  std::sort(currentValues.begin(), currentValues.end(), [](Value* a, Value* b) {
```
There won't be duplicate ids for `Value` objects. So sort and stable sort are the same here?

yiming0416 commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1522 @@
+  }
+
+  // Remove all nodes not in visited (other than input/outputs)
+  std::vector<Node*> toRemove;
```
The nodes to be removed may just be a small subset of all the nodes. So that may lead to over reserving?

zhxchen17 commented:
@albanD this pr is a bit over the 2000 lines limit and we're adding 1 logical component (graph) here. any opinion on the sanity check?

swolchok commented on test/cpp/nativert/test_graph.cpp:
```diff
@@ -0,0 +1,570 @@
+
+  {
+    // Check we shouldn't be able to remove a value that still has users
+    EXPECT_DEATH(
```
ditto EXPECT_THROW

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+    return kind_;
+  }
+
+  std::string classFqn() const {
+    CHECK(kind_ == Kind::CustomObj)
+        << "Only CustomObj type can have classFqn, got " << kind_;
+    return classFqn_;
+  }
```
why waste the memory for it for all other kinds then? instead, you could represent this class with std::variant internally and materialize the Kind on access

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+} // namespace
+
+bool operator==(const Type& left, const Type& right) {
+  return left.kind() == right.kind();
```
this doesn't compare the classFqn if necessary; migrating to std::variant as suggested in the header would get it fixed for free

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+    TORCH_CHECK_EQ(name_, this->name());
+  }
+
+  // Each Value should be uniquely created and managed by a Graph. It's an anti
+  // pattern to copy/move Value instances anyway.
```
how can copy/move be an anti-pattern when it's not allowed?

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+ public:
+  explicit Value(ValueId id, std::string name, Type t, Node* producer)
+      : name_(std::move(name)), id_(id), type_(t), producer_(producer) {
+    TORCH_CHECK_EQ(name_, this->name());
```
why do we need this check? name() returns name_. if there is some debugging reason we need it, how about TORCH_INTERNAL_ASSERT_DEBUG_ONLY?

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+  }
+
+  const Node* producer(bool resolve_folded = false) const {
+    return !resolve_folded && isFolded() ? nullptr : producer_;
```
I would parenthesize this for clarity -- `return (!resolve_folded && isFolded()) ? nullptr : producer_;`

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+    users_.clear();
+  }
+
+  // Fatals if the value is not a TensorList
```
no it doesn't, it TORCH_CHECKs, which throws an exception. I see several other similar comments like this below, please fix

malfet commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+
+class Type {
+ public:
+  enum Kind {
```
Please do not use enum in 2025! Only `enum class`

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+  void updateInputName(std::string_view oldName, std::string_view newName) {
+    for (auto& input : inputs_) {
+      if (input.name == oldName) {
+        input.name = newName;
```
is it an invariant that names are unique? can we terminate the loop after we find a match?

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+  // Return a bidirectional range over `const Value*`
+  // NOTE: this range can be invalidated by mutations to the graph.
+  auto outputs() const {
+    std::vector<const Value*> ret;
```
allocating a vector here seems inefficient. Since the return value here is fragile and can be invalidated by mutations anyway, couldn't we return an adapter class that holds a reference to outputNode_->inputs()?

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+  Node* insertionPoint() {
+    // This should never happen, since the last-most insertion point is the
+    // prim.Outputs node, not end().
+    CHECK(insertBefore_ != nodes_.end());
```
why CHECK instead of TORCH_CHECK?

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+  }
+
+  void setInsertionPoint(Node* n) {
+    CHECK(n != inputNode_) << "can't insert before prim.Input";
```
ditto

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+  // Node lifetime is managed by nodesOwner_, but the actual ordering is
+  // maintained intrusively using nodes_.
+  // This is to facilitate quick insertion before/after a given Node*.
+  std::vector<std::unique_ptr<Node>> nodesOwner_;
+  c10::IntrusiveList<Node> nodes_;
```
this doesn't strike me as a particularly efficient representation. It seems like either of the following would be better:
1) a vector that owns Nodes directly instead of by unique_ptr, and represent the ordering-maintaining linked list using array offsets instead of pointers, or
2) an owning std::list

I don't recall seeing an answer to @albanD's previous question about why we need c10::IntrusiveList.

yiming0416 replied:
@swolchok As discussed in https://github.com/pytorch/pytorch/pull/152754
> IntrusiveList is different from std::list because it allows us to allocate all the memory at once for a node, while std::list forces us to heap allocate twice for list entry and the node itself separately. In addition to memory efficiency and performance concerns, this also allows us to track nodes safely without passing around the list object (because intrusive list maintains a stronger invariant that expiration will always cause unlinking).

swolchok replied:
> std::list forces us to heap allocate twice for list entry and the node itself separately

I don't think this is true; replied on #152754.

> this also allows us to track nodes safely without passing around the list object (because intrusive list maintains a stronger invariant that expiration will always cause unlinking)

I am confused about the circumstances in which the InstrusiveList could be destroyed without `nodesOwner_` also being destroyed.

yiming0416 replied:
cc @zhxchen17

swolchok commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+  // constantSymIntValues_ is a subset of values_
+  std::unordered_map<ValueId, int> constantSymIntValues_;
+  // Output values of the graph, which is a subset of values_.
+  std::vector<std::variant<Value*, Constant>> userOutputs_;
```
this isn't particularly efficient; Constant is already a std::variant and ideally we would find a way to fuse the tags.

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+
+namespace {
+
+// Workaround for MSVC bug.
```
what bug? in what version(s) of MSVC?

yiming0416 replied:
`"std": ambiguous symbol`. I'll add to the comments

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+constexpr bool is_same_v = std::is_same_v<T, U>;
+
+bool isBlank(char n) {
+  return isspace(n) || n == '\n';
```
std::isspace('\n') is true already https://en.cppreference.com/w/c/string/byte/isspace

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+}
+
+Node* Graph::insertBefore(Node* toInsert, Node* insertionPoint) {
+  CHECK(insertionPoint != inputNode_) << "can't insert before prim.Input";
```
ditto TORCH_CHECK

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+  // number of elements, so it shouldn't be slow. This allows us to avoid a
+  // second datastructure for lookups.
+  // Drop a debug check here, just to make sure :)
+  TORCH_CHECK_LT(inputs_.size(), 1000);
```
did you want TORCH_INTERNAL_ASSERT_DEBUG_ONLY?

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+  // number of elements, so it shouldn't be slow. This allows us to avoid a
+  // second datastructure for lookups.
+  // Drop a debug check here, just to make sure :)
+  TORCH_CHECK_LT(attributes_.size(), 1000);
```
ditto

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+
+bool Graph::cleanupDeadNodes() {
+  std::unordered_set<const Node*> visited;
+  std::queue<const Node*> visitQueue;
```
if it's not important to do this breadth-first instead of depth-first, I would recommend using a std::vector as a stack instead of using a queue. (if it is important, I would recommend leaving a comment explaining why)

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+      continue;
+    }
+
+    std::vector<NamedArgument> clonedInputs;
```
missing reserve

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+    }
+
+    Node* newNode =
+        insertNode(std::string(n.target()), clonedInputs, n.metadata());
```
std::move(clonedInputs)

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+
+      std::visit(
+          [&](auto&& val) -> void {
+            // Workaround for MSVC bug
```
what bug in which version of MSVC?

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+[[maybe_unused]] inline constexpr bool AlwaysFalse = false;
+
+c10::IValue constantToIValue(const Constant& constant) {
+  // Workaround for MSVC bug
```
again, what bug in which version of MSVC?

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+        if constexpr (is_same_v<T, None>) {
+          return c10::IValue();
+        } else if constexpr (
+            is_same_v<T, int64_t> || is_same_v<T, vector<int64_t>> ||
+            is_same_v<T, double> || is_same_v<T, vector<double>> ||
+            is_same_v<T, string> || is_same_v<T, bool> ||
+            is_same_v<T, vector<bool>> || is_same_v<T, c10::ScalarType> ||
+            is_same_v<T, c10::MemoryFormat> || is_same_v<T, c10::Layout> ||
+            is_same_v<T, c10::Device> || is_same_v<T, vector<string>>) {
```
would `std::is_convertible_v<c10::IValue, T>` get this job done more succinctly, or do you actively want people to have to write out new types here?

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+    const T& list,
+    F formatter) {
+  if (encloseInSquareBrackets) {
+    out << "[";
```
nit: `'['`

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+    formatter(out, el);
+  }
+  if (encloseInSquareBrackets) {
+    out << "]";
```
ditto character literal

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+      // First time around; initialize our type sentinel with the first value.
+      // We will use this on subsequent iterations to check that all types are
+      // the same.
+      if (std::holds_alternative<int64_t>(val)) {
+        type = std::get<int64_t>(val);
+      } else if (std::holds_alternative<double>(val)) {
+        type = std::get<double>(val);
+      } else if (std::holds_alternative<bool>(val)) {
+        type = std::get<bool>(val);
```
why not use get_if?

swolchok commented on torch/nativert/graph/Graph.cpp:
```diff
@@ -0,0 +1,1527 @@
+
+/**
+ * Deserialization for graphs: parse the output produced by operator<<(Graph).
+ * This parser really only expects the exact output generated by well-formed
+ * Graph objects, so it is not very permissive and does not give good error
```
I see some cases of handwritten input internally (and don't really see cases of machine-generated input outside tests). Wouldn't it be better DevEx if we gave folks reasonable errors?

swolchok reviewed (changes requested):
how important is efficiency for this code? I'm worried about it, but maybe it's known not to matter?

yiming0416 commented:
> how important is efficiency for this code? I'm worried about it, but maybe it's known not to matter?
@swolchok Thank you for all the comments and suggestions!  At this point, our primary focus is on getting this open-sourced ASAP, and there are other ongoing efforts to improve efficiency in other parts of the runtime. 
I've addressed most of your comments. If possible, I'd like to defer a few of the remaining points to follow-up tasks.

yiming0416 commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,696 @@
+    return producer_;
+  }
+
+  void addUser(Node* node);
+  void eraseUser(Node* node);
```
@malfet  Pls see the comments about lifetime above.  We are also adding a README file for `NativeRT` in https://github.com/pytorch/pytorch/pull/154581

yiming0416 committed:
[nativert] Move Graph to pytorch core (#153994)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/153994

Move Graph to pytorch core.

Test Plan: buck2 test mode/opt caffe2/test/cpp/nativert:graph_test

Differential Revision: D74749418

yiming0416 commented on torch/nativert/graph/Graph.h:
```diff
@@ -0,0 +1,717 @@
+  /*implicit*/ Type(Kind kind) : kind_(kind) {}
+
+  // For CustomObj kind with classFqn
+  explicit Type(Kind kind, const std::string& classFqn)
+      : kind_(CustomObjData{classFqn}) {
+    TORCH_CHECK(kind == Kind::CustomObj);
```
@malfet  Here we want to increase the readability when people calling the constructor of `Type` explicitly to ensure only `CustomObj` can be passed in.

swolchok commented:
> splitting this PR into smaller ones. see stack #154532

do you need to close this one then?
